Cleaning yearly, quarterly and monthly date offsets

Pawel-Kranzberg opened the pull request:
- Ability to use `YearBegin` and `QuarterBegin` DateOffset classes as Period `freq`, along with offset aliases.
- Ability to use three-letter month short names as `month` / `startingMonth` arguments of year- / quarter-based offset classes (as already used in their `_from_name` class method).
- Expanded docstrings with, e.g., additional examples.
- Partial refactorization, as a step toward combining `QuarterOffset` and `YearOffset` `__init__` (mentioned as TO DO in the code).

- [ ] xref #38859
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

Pawel-Kranzberg committed:
Update offsets.pyx

Fixes to year, quarter and month offsets.

Pawel-Kranzberg committed:
Merge remote-tracking branch 'upstream/master'

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

jreback commented:
@Pawel-Kranzberg you don't need multiple PR
pls use the same one and just push new commits

jbrockmendel commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1838,12 +1838,12 @@ cdef class YearOffset(SingleConstructorOffset):
-
-    cdef readonly:
-        int month
+    cdef readonly int month
```
we generally use the patternw tih only a single `cdef`

```
cdef:
    public int _period_dtype_code
    readonly int month
    [...]
```

Pawel-Kranzberg replied:
@jbrockmendel  - Thanks, I've just fixed that. 
The problem with the failing test stems from the change of default month for BQuarterBegin and QuarterBegin classes, from 3 (March) to 1 (January). I believe the old values to be wrong, and leading to behaviours that the users might not expect, and inconsistency vs BQuarterEnd and QuarterEnd. I've for instance expected the first quarter to begin with January and end in March (that's already the case for BQuarterEnd and QuarterEnd), and not begin in March and end in May (the legacy BQuarterBegin and QuarterBegin do just that).
However, **the current tests expect the legacy (IMO incorrect) quarter spans. How should I go about it?**

Pawel-Kranzberg committed:
Update offsets.pyx

Minor formatting changes.

Pawel-Kranzberg committed:
Update offsets.pyx

Minor formatting changes.

Pawel-Kranzberg committed:
Add ability to provide months as short names

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Temporary unfixing of the default months for `QuarterBegin` and `BQuarterBegin`.

jreback commented:
if you are just refactoring then all tests should pass

please do any breaking change after this PR - it's very hard to tell what you are doing if these are conflated

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg commented:
@jreback - It's hard for me to do just refactoring when the final test is failing due to the class `QuarterBegin` having different default values depending on the creation method. See:
```
cdef class QuarterBegin(QuarterOffset):
    """
    DateOffset increments between Quarter start dates.
    startingMonth = 1 corresponds to dates like 1/01/2007, 4/01/2007, ...
    startingMonth = 2 corresponds to dates like 2/01/2007, 5/01/2007, ...
    startingMonth = 3 corresponds to dates like 3/01/2007, 6/01/2007, ...
    """
    _default_starting_month = 3
    _from_name_starting_month = 1
    _prefix = "QS"
    _day_opt = "start"
```
The test is as follows:
```
        sdate = datetime(1999, 12, 25)
        edate = datetime(2000, 1, 1)

        idx1 = date_range(start=sdate, end=edate, freq="QS")
        idx2 = date_range(
            start=sdate, end=edate, freq=offsets.QuarterBegin(startingMonth=1)
        )
>       assert len(idx1) == len(idx2)
```
Please advise me how to proceed. Should I, e.g., now create and process a new PR with only the default month changes? Or roll back refactoring of Quarter classes in the current PR?

jreback commented:
pls don't do refactoring and a bug fix in the same PR
one or the other or this will be impossible to merge

jreback commented:
be honest i don't know which bug this is supposedly fixing

Pawel-Kranzberg commented:
> be honest i don't know which bug this is supposedly fixing

@jreback - OK, I'd say it is a valid bug that a `QuarterBegin` object has different starting months, depending on if it was created using:

-  the offset alias. e.g., in `date_range(start=sdate, end=edate, freq="QS")` , with the resulting quarter having January as the starting month (`_from_name_starting_month = 1`), or
- directly, e.g. in `date_range(start=sdate, end=edate, freq=offsets.QuarterBegin()`, with March as the starting month (`_default_starting_month = 3`).

In addition I'd argue that default starting months for `QuarterBegin` and `BQuarterBegin` should be 1 instead of the current 3.

Pawel-Kranzberg commented:
Other breaking changes that I introduced allow using year, quarter and month starts, as well as business quarter and month starts, as `freq` in `date_range`.

jreback commented:
pls search the tracker pretty sure there are existing issues for this

we like targeted PRs
PRa that do too many things won't be merged

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg commented:
@jreback - Thank you for all the guidance and patience. This PR is now narrowed down to the following:

- Ability to use all year- and quarter-based offset classes as Period `freq`, along with offset aliases.
- Ability to use three-letter month short names as `month` / `startingMonth` arguments of year- / quarter-based offset classes (as already used in their `_from_name` class method).
- Expanded docstrings with, e.g., additional examples.
- Partial refactorization, as a step toward combining `QuarterOffset` and `YearOffset` __init__ (mentioned as TO DO in the code).

Is that scope acceptable?

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1903,30 +1913,48 @@ cdef class YearOffset(SingleConstructorOffset):
+    Timestamp('2021-01-01 05:01:15')
+    >>> ts + YearBegin(normalize = True)
+    Timestamp('2021-01-01 00:00:00')
+    >>> ts + YearBegin()._from_name('APR')
```
i wouldn't show internal methods

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1903,30 +1913,48 @@ cdef class YearOffset(SingleConstructorOffset):
-    Timestamp('2020-11-30 05:01:15')
+    >>> ts + YearEnd(normalize = True)
+    Timestamp('2020-12-31 00:00:00')
+    >>> ts + YearEnd()._from_name('AUG')
```
same

jreback reviewed (changes requested):
can you add some tests for construction from the str & anything else that is an enhancement. I am not exactly sure where this is useful. maybe some tests would show. Please also craft a whatsnew that shows from a user POV what has changed.

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

jreback commented:
@Pawel-Kranzberg I am still not sure what this PR is doing. seems you updated some doc-string & added some new fucntionaility which is not tested.

jbrockmendel commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1838,21 +1838,28 @@ cdef class YearOffset(SingleConstructorOffset):
-        int month
+    cdef:
+        readonly int month, _period_dtype_code
+        # public int _period_dtype_code
```
pls dont leave commented-out code

jbrockmendel commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1838,21 +1838,28 @@ cdef class YearOffset(SingleConstructorOffset):
 
-    # _default_month: int  # FIXME: python annotation here breaks things
+    _attributes = tuple(["n", "normalize", "month"])
+    _default_month = 1
```
if this cant be annotated, then please leave the comment

jbrockmendel commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1838,21 +1838,28 @@ cdef class YearOffset(SingleConstructorOffset):
-        self.month = month
+        if month is not None:
+            if isinstance(month, str):
+                self.month = MONTH_TO_CAL_NUM[month]
```
this is adding support for passing a month string?  thats not necessarily a bad idea, but would need to check perf impact

jbrockmendel commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -1903,30 +1913,44 @@ cdef class YearOffset(SingleConstructorOffset):
 
 
-cdef class BYearEnd(YearOffset):
+cdef class YearBegin(YearOffset):
```
are you re-ordering the classes in this file?  or is git just mangling the diff?

jbrockmendel commented:
pls start by reverting the re-ordering of classes.  that will make it much easier to see what is actually changed

jbrockmendel commented:
@Pawel-Kranzberg can you address comments and merge master

Pawel-Kranzberg commented:
> @Pawel-Kranzberg can you address comments and merge master

Yes, I will. I'm sorry, I've been very busy lately.

Pawel-Kranzberg committed:
Merge remote-tracking branch 'upstream/master'

Pawel-Kranzberg committed:
Merge pull request #2 from pandas-dev/master

Sync with master

Pawel-Kranzberg commented:
> This pull request is stale because it has been open for thirty days with no activity. Please update or respond to this comment if you're still interested in working on this.

I need it for now as a reference.
